kani-compiler/src/kani_middle/resolve.rs

```diff
@@ -0,0 +1,236 @@
+
+use std::collections::VecDeque;
+
+use rustc_hir::def::{DefKind, Res};
+use rustc_hir::def_id::{DefId, LocalDefId, CRATE_DEF_INDEX};
+use rustc_hir::ItemKind;
+use rustc_middle::ty::TyCtxt;
+
+/// Attempts to resolve a path (in the form of a string) to a `DefId`. The
+/// current module is provided as an argument in order to resolve relative
+/// paths.
+pub fn resolve_path(tcx: TyCtxt, current_module: LocalDefId, path_str: &str) -> Option<DefId> {
+    let path = to_path(tcx, current_module, path_str)?;
+    match &path.base {
+        Base::ExternPrelude => resolve_external(tcx, path.segments),
+        Base::LocalModule { id, may_be_external_path } => {
+            // Try to resolve it as a relative path first; if this fails and the
+            // path might be external (it wasn't qualified with `self`, etc.)
+            // and the current module does not have a submodule with the same
+            // first segment, try resolving it as an external path.
+            resolve_relative(tcx, *id, path.segments.clone()).or_else(|| {
+                if *may_be_external_path
+                    && !has_submodule_with_name(tcx, current_module, path.segments.front()?)
+                {
+                    resolve_external(tcx, path.segments)
+                } else {
+                    None
+                }
+            })
+        }
+    }
+}
+
+/// The segments of a path.
+type Segments = VecDeque<String>;
+
+/// The "starting point" for a path.
+#[derive(Debug)]
+enum Base {
+    /// Indicates an external path.
+    ExternPrelude,
+    /// Indicates a path that may be local (and must be local if
+    /// `may_be_external_path` is false) and should be resolved relative to the
+    /// module identified by `id`.
+    LocalModule { id: LocalDefId, may_be_external_path: bool },
+}
+
+/// A path consisting of a starting point and a bunch of segments.
+#[derive(Debug)]
+struct Path {
+    base: Base,
+    segments: Segments,
+}
+
+impl Path {
+    fn new(base: Base, segments: Segments) -> Self {
+        Path { base, segments }
+    }
+}
+
+/// Takes a string representation of a path and turns it into a `Path` data
+/// structure, resolving qualifiers (like `crate`, etc.) along the way.
+fn to_path(tcx: TyCtxt, current_module: LocalDefId, name: &str) -> Option<Path> {
+    const CRATE: &str = "crate";
+    const ROOT: &str = "{{root}}";
+    const SELF: &str = "self";
+    const SUPER: &str = "super";
+
+    // Split the string into segments separated by `::`.
+    let mut segments: Segments = name.split("::").map(str::to_string).collect();
+    if segments.is_empty() {
+        return Some(Path::new(
+            Base::LocalModule { id: current_module, may_be_external_path: false },
+            segments,
+        ));
+    }
+
+    // Resolve qualifiers `crate`, initial `::`, and `self`. The qualifier
+    // `self` may be followed be `super` (handled below).
+    let mut may_be_external_path = true;
```

Comment:
For readability, I would recommend inspecting the first element and assigning the proper value here. Something like:
```
let may_be_external_path = !matches!(first, CRATE, SELF, SUPER);
```

Reply:
This is cleaner! Done